zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+- ZAP PR: [zulip/zulip-architecture#2](https://github.com/zulip/zulip-architecture/pull/2)
+
+# Summary
+
+Proposal to add end-to-end encryption for mobile push notifications.
+
+# Motivation
+
+Currently, the Zulip Push Notification Forwarding Service receives
+(but does not store or log) the contents of individual mobile push notifications.
+
+Adding end-to-end encryption for push notifications will ensure that the
+actual message content does not need to be readable by the push notification
+bouncer (or Google/Apple's servers, for that matter).
+
+# Detailed design
+
+## High-Level Plan
+
+- Client checks the server if it supports E2E encryption. If it doesn't,
+  use the legacy push notification registration flow.
+
+- Client makes an API call to the server to register itself for push notifications,
+  receives a shared secret key in response.
+
+- While sending notification, the server encrypts the notification content
+  and sends it to push bouncer. Push bouncer forwards it to FCM/APNs.
+
+- Client receives the notification via FCM/APNs and decrypts the content
+  using the shared secret key.
+
+## Data model
+
+- On the client:
+  ```
+  model Account:
+    …
+    push_account_id: RandomID[32]  # Generated by client; unique here.
+    last_acked_push_token: str  # The last push token the server acknowledged receipt of from this client.
+    push_key: SymmetricKey  # Shared key between client and server.
+  ```
+  ```
+  // A helper data structure
+  struct PushRegistration:
+    token_kind: "apns" | "fcm"
+    token: str
+    timestamp: datetime
+  ```
+- On the server, add a new table (parallel to `PushDeviceToken`):
+  ```
+  class zerver.models.push_notifications.PushDevice:
+    user: ForeignKey[UserProfile]
+    encrypted_push_registration: Ciphertext # A PushRegistration, encrypted with bouncer_key.
+    push_account_id: RandomID[32]  # Generated by client; need not be unique.
+    bouncer_device_id: RandomID[32] | None  # Generated by bouncer as its unique ID for this record. Null until bouncer has confirmed the registration.
+    push_key: SymmetricKey # Shared key between client and server.
+  ```
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
```

Comment:
TODO: There's a comment thread in the doc regarding "how the import tool should handle this table", I need to document it here.

Reply:
Done in second commit.

Reply:
Actually, we don't need to import `PushDevice`.

See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2063815370